### ISSUE TITLE
Issue with encoding

### ISSUE THREAD

**Issue by dmacd4484** [rOpenGov/pxweb#217]
I was wondering if I could ask your help. I have been using your “pxweb” R package to access the API at the Slovenian Statistical Agency (SORS). Thank you for developing and maintaining this very clear and straightforward API interface. 
 
However, I have run into a problem. For some queries, the code suggested by pxweb_interactive() is not accepted by pxweb_get(). This seems to be an issue with some of the character encoding. I have included a code example to illustrate the issue (see below). 
 
In the first data series of the code example (Long-term unemployment rate) this process works well, and the code acts as expected. That is, it accesses the API and saves the data into a data frame.
 
In the second series, there seems to be an issue with the encoding, as the Č character is not being recognised. I have attempted to adjust the coding and have even entered in the Unicode characters directly, with no success.
 
Would you have any insight into how I might be able to fix this issue. Based on the traceback, the code throws an error on an assertion, and so I am not sure if the query would return anything if it were sent to the API itself.
 
 
# Example of the error:
 
### PXWEB query 
pxweb_query_list <- 
  list("KOHEZIJSKA REGIJA"=c("0"),
       "DRŽAVA ROJSTVA"=c("0"),
       "SPOL"=c("0"),
  "ČETRTLETJE"=c("2008Q1","2008Q2","2008Q3","2008Q4","2009Q1","2009Q2","2009Q3","2009Q4","2010Q1","2010Q2","2010Q3","2010Q4","2011Q1","2011Q2","2011Q3","2011Q4","2012Q1","2012Q2","2012Q3","2012Q4","2013Q1","2013Q2","2013Q3","2013Q4","2014Q1","2014Q2","2014Q3","2014Q4","2015Q1","2015Q2","2015Q3","2015Q4","2016Q1","2016Q2","2016Q3","2016Q4","2017Q1","2017Q2","2017Q3","2017Q4","2018Q1","2018Q2","2018Q3","2018Q4","2019Q1","2019Q2","2019Q3","2019Q4","2020Q1","2020Q2","2020Q3"),
       "MERITVE"=c("2000"))
 
### Download data 
px_data <- 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px",
            query = pxweb_query_list)
 
**Error in pxweb_add_metadata_to_query(pxq, pxmd) : 
  Assertion on 'pxweb_query_variable_code' failed: Must be element of set {'KOHEZIJSKA REGIJA','DRŽAVA ROJSTVA','SPOL','CETRTLETJE','MERITVE'}, but is 'CETRTLETJE'.**

# Code:
 
library(pxweb)
 
 
################################################################################
################################################################################
 
### Code to call interactive interface to generate PXWEB code
url <- "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762116S.px"
d <- pxweb_interactive(url)
### PXWEB query
pxweb_query_list <- 
  list("KAZALNIKI"=c("0"),
       "LETO"=c("2008","2009","2010","2011","2012","2013","2014","2015","2016","2017","2018","2019"))
 
### Download data 
px_data <- 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762116S.px",
            query = pxweb_query_list)
 
### Convert to data.frame 
px_data <- as.data.frame(px_data, column.name.type = "text", variable.value.type = "text")
 
################################################################################
################################################################################
 
### Code to call interactive interface to generate PXWEB code
url <- "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px"
d <- pxweb_interactive(url)
### PXWEB query 
pxweb_query_list <- 
  list("KOHEZIJSKA REGIJA"=c("0"),
       "DRŽAVA ROJSTVA"=c("0"),
       "SPOL"=c("0"),
       "ČETRTLETJE"=c("2008Q1","2008Q2","2008Q3","2008Q4","2009Q1","2009Q2","2009Q3","2009Q4","2010Q1","2010Q2","2010Q3","2010Q4","2011Q1","2011Q2","2011Q3","2011Q4","2012Q1","2012Q2","2012Q3","2012Q4","2013Q1","2013Q2","2013Q3","2013Q4","2014Q1","2014Q2","2014Q3","2014Q4","2015Q1","2015Q2","2015Q3","2015Q4","2016Q1","2016Q2","2016Q3","2016Q4","2017Q1","2017Q2","2017Q3","2017Q4","2018Q1","2018Q2","2018Q3","2018Q4","2019Q1","2019Q2","2019Q3","2019Q4","2020Q1","2020Q2","2020Q3"),
       "MERITVE"=c("2000"))
 
### Download data 
px_data <- 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px",
            query = pxweb_query_list)
 
### Convert to data.frame 
px_data <- as.data.frame(px_data, column.name.type = "text", variable.value.type = "text")

**Comment by MansMeg** [comment]
Hi! 

Thanks for this issue. I can get this to work (see below) so I think it is probably an encoding issue. What OS are you using? Windows? Could you try to save the Rscript as UTF8? I think that is the default of all PXWEB APIs.

```
library(pxweb)
pxweb_query_list <-
 
  list("KOHEZIJSKA REGIJA"=c("0"),
       
       "DRŽAVA ROJSTVA"=c("0"),
       
       "SPOL"=c("0"),
       
       "ČETRTLETJE"=c("2008Q1","2008Q2","2008Q3","2008Q4","2009Q1","2009Q2","2009Q3","2009Q4","2010Q1","2010Q2","2010Q3","2010Q4","2011Q1","2011Q2","2011Q3","2011Q4","2012Q1","2012Q2","2012Q3","2012Q4","2013Q1","2013Q2","2013Q3","2013Q4","2014Q1","2014Q2","2014Q3","2014Q4","2015Q1","2015Q2","2015Q3","2015Q4","2016Q1","2016Q2","2016Q3","2016Q4","2017Q1","2017Q2","2017Q3","2017Q4","2018Q1","2018Q2","2018Q3","2018Q4","2019Q1","2019Q2","2019Q3","2019Q4","2020Q1","2020Q2","2020Q3"),
       
       "MERITVE"=c("2000"))



# Download data

> px_data <-
 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px",
           
            query = pxweb_query_list)

> px_data
PXWEB DATA
With 6 variables and 51 observations.
```

**Comment by dmacd4484** [comment]
Thanks for the quick reply. Yes, I am using Windows. I had saved the file as UTF-8, and it still does not work. Though it had originally be saved at UTF-8-BOM.I am not sure of that is an issue.

**Comment by MansMeg** [comment]
Ok. I do not really know how to handle encodings in a good way in Windows. 

Maybe this would work:
```
Sys.setlocale(locale="UTF-8") 
```

Otherwise, can you store the call as a JSON query and store it as a separate JSON-file (with UTF8)?

**Comment by MansMeg** [comment]
I have now been able to implement your error as a test case. It seem to be an encoding issue with Windows.

https://github.com/rOpenGov/pxweb/pull/220/checks?check_run_id=1973116844

The Windows build get the same error as you have.

**Comment by dmacd4484** [comment]
Ah that is good to hear. It means I must not be crazy :)

So the issue is stemming from Windows itself rather than anything that I can adjust with the code or with R/Rstudio?

I am trying to run this code on a computer where I don't have Admin rights, but I am asking my IT department if they can let me change the encoding to UTF-8. Apparently Windows 10 has this as a beta option in the System -> Region settings. I will report back if that work.


Just as a reference, this website seems to describe the issue clearly:
https://developer.r-project.org/Blog/public/2020/05/02/utf-8-support-on-windows/

**Comment by MansMeg** [comment]
Yes. It seems to be windows. I think there must be an easier way than to switch the region settings. I would expect the problems to be solvable without this as well. The problem is that I have no windows machine to do the debugging.

If you would do the debugging I would be happy to help and update the package. The problem is just to find out exactly where the problem comes in. PXWEB are built on utf8 so there must be somewhere where the query is converted incorrectly.

**Comment by dmacd4484** [comment]
Happy to do the debugging. 

Here is the traceback() of the error on my machine:

Error in pxweb_add_metadata_to_query(pxq, pxmd) : 
  Assertion on 'pxweb_query_variable_code' failed: Must be element of set {'KOHEZIJSKA REGIJA','DRŽAVA ROJSTVA','SPOL','CETRTLETJE','MERITVE'}, but is 'CETRTLETJE'.

> traceback()
7: stop(simpleError(sprintf(msg, ...), call.))
6: mstop("Assertion on '%s' failed: %s.", var.name, res, call. = sys.call(-2L))
5: makeAssertion(x, res, .var.name, add)
4: checkmate::assert_choice(pxweb_query_variable_code, choices = pxweb_metadata_variables)
3: pxweb_add_metadata_to_query(pxq, pxmd)
2: pxweb_advanced_get(url = url, query = query, verbose = verbose)
1: pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px", 
       query = pxweb_query_list)

**Comment by MansMeg** [comment]
So it seems like it is pxweb_add_metadata_to_query(pxq, pxmd) that is essentially adding metadata to the query where the problem shows. 

I cannot, unfortunately, give you more information or debug this over github based on your output.

If you have the possibility: 
1. Fork this repo and then clone the repository and build the package locally on your computer. 
2. Use (for example) `debugonce("pxweb_get")` to see what exactly happens in all steps. I think, somewhere, ČETRTLETJE seem to become CETRTLETJE. 
3. See if you can fix that so it works on your computer by changing that small part of the code where the error occurs.
4. Commit the fix to your forked repo and do a pull request to this repo and the test will run and we will see if it fixes the error.

Sorry for this hassle.

**Comment by dmacd4484** [comment]
Ok sounds like a good plan. I will try this and see how far I can get.

Thanks for the help.

**Comment by MansMeg** [comment]
Great! Let me know if you run into a problem somewhere. And thanks for contributing!

**Comment by dmacd4484** [comment]
Hi Mans,

I found a fix, but it did not involve having to make any adjustments to the **pxweb**. 

In attempting to debug the problem, I  realized that the issue was in the arguments being passed. For example, in R or Rstudio the UTF-8 coding is not preserved during assignment in Windows. For instance:
`x <- "ČETRTLETJE"`
`x`
[1] "CETRTLETJE"

So when I pass the **pxweb_query_list** provided by **pxweb_interactive()**, while there is nothing wrong with the query, the encoding becomes garbled.

To fix this, I explicitly referenced the offending character using the escape sequence. 

This seems to work. It is an ugly fix. But it is a fix nonetheless.

Thanks a lot for your help in finding the error, and also for the extremely useful package.

Duncan 

## Query that gives error:
#### PXWEB query 
pxweb_query_list <- 
```
  list("KOHEZIJSKA REGIJA"=c("0"),
       "DRŽAVA ROJSTVA"=c("0"),
       "SPOL"=c("0")", 
"ČETRTLETJE"=c("2008Q1","2008Q2","2008Q3","2008Q4","2009Q1","2009Q2","2009Q3","2009Q4","2010Q1","2010Q2","2010Q3","2010Q4","2011Q1","2011Q2","2011Q3","2011Q4","2012Q1","2012Q2","2012Q3","2012Q4","2013Q1","2013Q2","2013Q3","2013Q4","2014Q1","2014Q2","2014Q3","2014Q4","2015Q1","2015Q2","2015Q3","2015Q4","2016Q1","2016Q2","2016Q3","2016Q4","2017Q1","2017Q2","2017Q3","2017Q4","2018Q1","2018Q2","2018Q3","2018Q4","2019Q1","2019Q2","2019Q3","2019Q4","2020Q1","2020Q2","2020Q3"),
       "MERITVE"=c("2000"))
```

#### Download data 
```
px_data <- 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px",
            query = pxweb_query_list)
```

## Fix for Windows:

#### PXWEB query 
```
pxweb_query_list <- 
  list("KOHEZIJSKA REGIJA"=c("0"),
       "DRŽAVA ROJSTVA"=c("0"),
       "SPOL"=c("0"),
       "MERITVE"=c("2000"))
```

#### Explicit encoding ČETRTLETJE
```
fixed.name <- paste("\U010C", "ETRTLETJE", sep = "")
years=c("2008Q1","2008Q2","2008Q3","2008Q4","2009Q1","2009Q2","2009Q3","2009Q4","2010Q1","2010Q2","2010Q3" ,"2010Q4","2011Q1","2011Q2","2011Q3","2011Q4","2012Q1","2012Q2","2012Q3","2012Q4","2013Q1","2013Q2","2013Q3","2013Q4","2014Q1","2014Q2","2014Q3","2014Q4","2015Q1","2015Q2","2015Q3","2015Q4","2016Q1","2016Q2","2016Q3","2016Q4","2017Q1","2017Q2","2017Q3","2017Q4","2018Q1","2018Q2","2018Q3","2018Q4","2019Q1","2019Q2","2019Q3","2019Q4","2020Q1","2020Q2","2020Q3")
pxweb_query_list[[fixed.name]] <- years
```

#### Download data 
```
px_data <- 
  pxweb_get(url = "https://pxweb.stat.si:443/SiStatData/api/v1/en/Data/0762002S.px",
            query = pxweb_query_list)
```

**Comment by MansMeg** [comment]
That's great!

If you have the possibility, it would be great if you could add this as a solution in:
https://github.com/rOpenGov/pxweb/blob/master/TROUBLESHOOTING.md

Then do a PR and also add yourself as a contributor to the package here:
https://github.com/rOpenGov/pxweb/blob/master/DESCRIPTION

I think you solved an important issue with the package so I think it would be fair to add you as a contributor.

**Comment by dmacd4484** [comment]
OK happy to add to the Troubleshooting page. As this issue is outside of the package itself, I don't know if I am a contributor to the package. I will add my name on the basis of adding the workaround description to the Troubleshooting page.

**Comment by MansMeg** [comment]
Yes, that was my idea as well. Looking forward!

**Comment by sakaiando** [comment]
I got the same error when I tried to download latvia data. I could download the data via interactive mode, so it seems the same issue is behind the error.

```
pxq <- 
  list("Gads/Ceturksnis"=c("*"),
       "Izglitibas limenis"=c("*"),
       "Ekonomiskas aktivitates statuss"=c("*"),
       "Raditaji"=c("*"),
       "Dzimums"=c("*"))
pxweb_get(url = "http://data1.csb.gov.lv/api/v1/en/sociala/nodarb/aktivitate/isterm/NB030c.px",
          query = pxq) %>%
  as.data.frame(column.name.type = "text", variable.value.type = "text") ->
  df
```

> Error in pxweb_add_metadata_to_query(pxq, pxmd) : 
>   Assertion on 'pxweb_query_variable_code' failed: Must be element of set {'Gads/Ceturksnis','Izglitibas limenis','Ekonomiskas aktivitates statuss','Dzimums','Raditaji'}, but is 'Izglitibas limenis'.


I tried the same trick (as below) but it seems the attempt is not successful. I would appreciate it if you could advise...

```
rm(list = ls())
cat("/014") 

library(tidyverse)
library(pxweb)

v1 <- 'Gads/Ceturksnis'
v2 <- paste("\U010C", "Izglitibas limenis", sep = "")
v3 <- paste("\U010C", "Ekonomiskas aktivitates statuss", sep = "")
v4 <- paste("\U010C", "Raditaji", sep = "")
v5 <- paste("\U010C", "Dzimums", sep = "")

pxq <- vector(mode="list",length=0)
pxq[[v1]] <- c('*')
pxq[[v2]] <- c('*')
pxq[[v3]] <- c('*')
pxq[[v4]] <- c('*')
pxq[[v5]] <- c('*')

pxweb_get(url = "http://data1.csb.gov.lv/api/v1/en/sociala/nodarb/aktivitate/isterm/NB030c.px",
          query = pxq) %>%
  as.data.frame(column.name.type = "text", variable.value.type = "text") ->
  df
```

> Error in pxweb_add_metadata_to_query(pxq, pxmd) : 
>   Assertion on 'pxweb_query_variable_code' failed: Must be element of set {'Gads/Ceturksnis','Izglitibas limenis','Ekonomiskas aktivitates statuss','Dzimums','Raditaji'}, but is 'C Izglitibas limenis'.

The session info is

> R version 3.6.3 (2020-02-29)
> Platform: x86_64-w64-mingw32/x64 (64-bit)
> Running under: Windows 10 x64 (build 18363)
> 
> Matrix products: default
> 
> locale:
> [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
> [3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
> [5] LC_TIME=English_United States.1252    
> 
> attached base packages:
> [1] stats     graphics  grDevices datasets  utils     methods   base     
> 
> other attached packages:
>  [1] pxweb_0.10.2    forcats_0.5.0   stringr_1.4.0   dplyr_0.8.5     purrr_0.3.3     readr_1.3.1    
>  [7] tidyr_1.0.2     tibble_2.1.3    ggplot2_3.3.0   tidyverse_1.3.0
> 
> loaded via a namespace (and not attached):
>  [1] tidyselect_1.0.0 xfun_0.12        haven_2.2.0      lattice_0.20-40  colorspace_1.4-1 vctrs_0.2.4     
>  [7] generics_0.0.2   htmltools_0.5.0  yaml_2.2.1       utf8_1.1.4       rlang_0.4.5      pillar_1.4.3    
> [13] glue_1.3.2       withr_2.1.2      DBI_1.1.0        dbplyr_1.4.2     modelr_0.1.6     readxl_1.3.1    
> [19] lifecycle_0.2.0  munsell_0.5.0    gtable_0.3.0     cellranger_1.1.0 rvest_0.3.5      evaluate_0.14   
> [25] knitr_1.28       curl_4.3         fansi_0.4.1      broom_0.5.5      Rcpp_1.0.3       checkmate_2.0.0 
> [31] scales_1.1.0     backports_1.1.5  jsonlite_1.6.1   fs_1.3.2         hms_0.5.3        digest_0.6.25   
> [37] stringi_1.4.6    grid_3.6.3       cli_2.0.2        tools_3.6.3      magrittr_1.5     crayon_1.3.4    
> [43] pkgconfig_2.0.3  xml2_1.2.5       reprex_0.3.0     lubridate_1.7.4  assertthat_0.2.1 rmarkdown_2.1   
> [49] httr_1.4.1       rstudioapi_0.11  R6_2.4.1         nlme_3.1-145     compiler_3.6.3  

Many thanks in advance.

**Comment by MansMeg** [comment]
It seems like you add a C to the name as well?

> but is 'C Izglitibas limenis'.

**Comment by sakaiando** [comment]
Thanks @MansMeg,

I didn't add C myself, but it seems paste("\U010C", "Izglitibas limenis", sep = "") generates C in front of "Izglitibas limenis".

I guess \U010C works for Slovenia data but needs to be replaced by something in the case of Latvia data.

Unfortunately, I am not familiar with encoding things, so if you or @dmacd4484 know a quick fix, please let me know...

**Comment by MansMeg** [comment]
So the problem is that some of your letters are special. I guess the problem here is https://en.m.wiktionary.org/wiki/ī and hence, you need to find the UTF8 encoding for this character (see the link). Then you add that character encoding directly (as above). You can try and see if R print this correct first.

**Comment by dmacd4484** [comment]
Hi @sakaiando, 

@MansMeg 's suggestion is spot on. You need to find out what special characters you need to replace in your specific queries. An easy way to do that is to put the px link into your browser and see what the actual codes are. 

In your case the link is:
http://data1.csb.gov.lv/api/v1/en/sociala/nodarb/aktivitate/isterm/NB030c.px 

**And your codes are:**
Gads/Ceturksnis
Izglītības līmenis
Ekonomiskās aktivitātes statuss
Dzimums
Rādītāji

So it seems that you have a number of vowels to replace in three of these codes. You can then take this text to a UTF converter such as [this on]( https://r12a.github.io/app-conversion/ ) where you can see the relevant escape codes and where to place them. In the link shown, just look at the Java or Ruby sections of the page after you enter your text and convert it.

**For example, the manual conversion process would involve steps like this:**
**Original :** Izglītības līmenis
**JS/Java/C:** Izgl\u{12B}t\u{12B}bas l\u{12B}menis
**R:** v2 <- paste("Izgl", "\U12B", "t", "\U12B" , "bas ", "\U12B", "menis", sep = "")

Hopefully this will work for you

**Comment by MansMeg** [comment]
Thank you @dmacd4484 , we should add this to the troubleshooting document since this seem to be a common problem for Windows users.

**Comment by sakaiando** [comment]
Thanks @MansMeg and @dmacd4484, the code below worked!

```
rm(list = ls())
cat("/014") 

library(tidyverse)
library(pxweb)

v1 <- 'Gads/Ceturksnis'
v2 <- paste("Izgl", "\U12B", "t", "\U12B" , "bas l", "\U12B", "menis", sep = "")
v3 <- paste("Ekonomisk","\u101","s aktivit","\u101","tes statuss", sep = "")
v4 <- 'Dzimums'
v5 <- paste("R","\u101","d","\u12B","t","\u101","ji", sep="")

pxq <- vector(mode="list",length=0)
pxq[[v1]] <- c('*')
pxq[[v2]] <- c('*')
pxq[[v3]] <- c('*')
pxq[[v4]] <- c('*')
pxq[[v5]] <- c('*')

pxweb_get(url = "http://data1.csb.gov.lv/api/v1/en/sociala/nodarb/aktivitate/isterm/NB030c.px",
          query = pxq) %>%
  as.data.frame(column.name.type = "text", variable.value.type = "text") ->
  df
```

> NAs introduced by coercion

**Comment by MansMeg** [comment]
Now fixed in version 0.11
